flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SourceReaderStreamTaskTest.java

```diff
@@ -18,71 +18,191 @@
+import org.apache.flink.runtime.state.TestTaskStateManager;
 import org.apache.flink.streaming.api.graph.StreamConfig;
 import org.apache.flink.streaming.api.operators.SourceReaderOperator;
 import org.apache.flink.streaming.runtime.io.InputStatus;
 import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
-import org.apache.flink.streaming.util.TestHarnessUtil;
+import org.apache.flink.util.ExceptionUtils;
 
 import org.junit.Test;
 
+import java.util.Iterator;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+
+import scala.concurrent.duration.Deadline;
+import scala.concurrent.duration.FiniteDuration;
+
+import static org.apache.flink.streaming.util.TestHarnessUtil.assertOutputEquals;
+import static org.apache.flink.util.Preconditions.checkState;
+import static org.junit.Assert.assertEquals;
 
 /**
  * Tests for verifying that the {@link SourceReaderOperator} as a task input can be integrated
  * well with {@link org.apache.flink.streaming.runtime.io.StreamOneInputProcessor}.
  */
 public class SourceReaderStreamTaskTest {
-
+	/**
+	 * Tests that the stream operator can snapshot and restore the operator state of chained
+	 * operators.
+	 */
 	@Test
-	public void testSourceOutputCorrectness() throws Exception {
+	public void testSnapshotAndRestore() throws Exception {
 		final int numRecords = 10;
+
+		TaskStateSnapshot taskStateSnapshot = executeAndWaitForCheckpoint(
+			numRecords,
+			1,
+			IntStream.range(0, numRecords),
+			Optional.empty());
+
+		executeAndWaitForCheckpoint(
+			numRecords,
+			2,
+			IntStream.range(numRecords, 2 * numRecords),
+			Optional.of(taskStateSnapshot));
+	}
+
+	private TaskStateSnapshot executeAndWaitForCheckpoint(
+			int numRecords,
+			long checkpointId,
+			IntStream expectedOutputStream,
+			Optional<TaskStateSnapshot> initialSnapshot) throws Exception {
+
+		final LinkedBlockingQueue<Object> expectedOutput = new LinkedBlockingQueue<>();
+		expectedOutputStream.forEach(record -> expectedOutput.add(new StreamRecord<>(record)));
+		CheckpointOptions checkpointOptions = CheckpointOptions.forCheckpointWithDefaultLocation();
+		expectedOutput.add(new CheckpointBarrier(checkpointId, checkpointId, checkpointOptions));
+
+		final Deadline deadline = new FiniteDuration(30, TimeUnit.SECONDS).fromNow();
+		final StreamTaskTestHarness<Integer> testHarness = createTestHarness(numRecords);
+		if (initialSnapshot.isPresent()) {
+			testHarness.setTaskStateSnapshot(checkpointId, initialSnapshot.get());
+		}
+
+		TestTaskStateManager taskStateManager = testHarness.taskStateManager;
+		OneShotLatch waitForAcknowledgeLatch = new OneShotLatch();
+
+		taskStateManager.setWaitForReportLatch(waitForAcknowledgeLatch);
+
+		testHarness.invoke();
+		testHarness.waitForTaskRunning(deadline.timeLeft().toMillis());
```

Comment:
Nit: Actually the passed timeout argument is ignored inside implementation of `testHarness.waitForTaskRunning`.
In addition why not pass 30000 directly to avoid the transformation of `deadline` here and below usage.

Reply:
Ohh, that's lovely. I will drop the timeout parameter in that case. Thanks for spotting this.

PS, I copied the base test structure including the `deadline` from some other test, that's why I didn't notice that indeed it would be better to have a constant "TIMEOUT" (assuming that it was honoured)

Reply:
Thanks for the fix.